README.md
# aks-nsg-ingress-ban-ip
Container: Ban malicious IP addresses in Azure NSG by AKS ingress logs.

Current limitation: Only NGINX Ingress controller log format is parsed (but feel free to extend the `eh-logs.js`).

NSG aka Network Security Group aka Firewall ;-)

This is a ready-to-use implementation using the 
[azure-nsg-ban-ips NPM package](https://www.npmjs.com/package/azure-nsg-ban-ips).

- The container subscribes to EventHub to read out the NGINX Ingress logs
- Malicious IPs are identified by lot of errors, caused by attacks they try
- These IPs are banned for some days in the firewall of the Kubernetes (aka NSG)

IMPORTANT: Project status is still EXPERIMENTAL!
 
# Who will be baned?

The Container/Pod will listen for NginX Ingress Controller access logs (or any similar format logs). Attackers try lot of things in a short time (scripted) to find backdoors, misconfigurations or security holes.

Attackers IP addresses are banned 
1. if they do more HTTP errors in a short time*
   - errors: status code > 399
   - default ERROR_THRESHOLD setting is 50
2. if they try to access URL paths containing NOGO_REQUESTS
   - default NOGO_REQUESTS pattern list is:
     "phpmyadmin,etc/passwd,wp-file-manager,phpunit" (feel free to extend them)
   - default NOGOS_THRESHOLD is 3 (# requests before banned) 

Blocking rules are created near realtime, if these rules are violated.

*) "Short time" means: Every minute the violation counter will be decreased by one. 
(attacker scripts do hundreds of attacks in seconds.)

# Set Up

... it's easier than it may look ;-)

If you don't have a Linux with a Bash, I recommend to use the 
Azure Cloud shell in your Browser: Click icon in Azure Portal top menu bar.

## Stream App Insights NGINGX logs to an Event Hub 
(... and other container logs, too)
 
You need to create an EventHub and configure your LogAnalytics to
stream the container logs there.

Just change the resource names to fit yours:
```sh
#!/bin/bash
# resource names:
export LOG_WORKSPACE_RESOURCE_GRP="my-group"
export LOG_WORKSPACE="my-log-analytics"
export EH_RESOURCE_GRP="my-group" # may be the same as above, you decide
export EVENT_HUB_NAMESPACE="my-log-eventhub-ns"
export LOGS_EVENT_HUB="my-container-logs"
export LOCATION="West Europe"
```

Login to Azure with `az login` and then
set up everything, just by running the following commands:


```sh
#!/bin/bash
# create resource group 
az group create --name "$EH_RESOURCE_GRP" --location "$LOCATION" 

# Event Hub Namespace (check if you perhaps need sku= Standard)
az eventhubs namespace create \
  --name "$EVENT_HUB_NAMESPACE" \
  --resource-group "$EH_RESOURCE_GRP" \
  --location "$LOCATION" \
  --capacity 1  \
  --enable-auto-inflate true \
  --enable-trusted-service-access true \
  --maximum-throughput-units 4 \
  --sku Basic

# create Event Hub in Namespace
az eventhubs eventhub create \
  --name "$LOGS_EVENT_HUB" \
  --namespace-name "$EVENT_HUB_NAMESPACE" \
  --resource-group  "$EH_RESOURCE_GRP"

# extract Event Hub ID
LOG_EH_ID=$(az eventhubs eventhub show --name $LOGS_EVENT_HUB --namespace-name "$EVENT_HUB_NAMESPACE" -g "$EH_RESOURCE_GRP" --query id -o tsv)

az eventhubs eventhub authorization-rule create \
  --eventhub-name $LOGS_EVENT_HUB \
  --name "aks-nsg-ingress-ban-ip" \
  --namespace-name "$EVENT_HUB_NAMESPACE" \
  --resource-group "$EH_RESOURCE_GRP" \
  --rights Listen

export EH_KEY=$(az eventhubs eventhub authorization-rule keys list --resource-group "$EH_RESOURCE_GRP" --namespace-name "$EVENT_HUB_NAMESPACE" --eventhub-name  "$LOGS_EVENT_HUB" --name "aks-nsg-ingress-ban-ip" --query primaryKey -o tsv)

# configure log analytics to stream container logs to Event Hub
az monitor log-analytics workspace data-export create \
  --name "eos-dev-export-containerlogs" \
  --workspace-name "$LOG_WORKSPACE" \
  --resource-group "$EH_RESOURCE_GRP" \
  --tables ContainerLog --destination "$LOG_EH_ID"

echo "Event Hub Key: $EH_KEY"
```

You need the key in the container configuration...

## Configure and Start "aks-nsg-ingress-ban-ip" Container in AKS

You don't need to clone this repo and build the container yourself, 
just copy and apply the [run-kubernetes-pod.yaml](run-kubernetes-pod.yaml).

Before you run the pod, you need these configuration items:
1. `AAD_ID` = the ID of your Azure Active Directory (where we need to login)
2. `SP_ID` = a Service Principal ID, so a technical account you need to
    prepare in AAD, see 
    [how to create a SP](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)
3. `SP_KEY` = the secret key for the Service Principal
4. `NSG` = name of the Network Security Group (aka Firewall) which need 
   to be configured. For an AKS this lives in the [Node Resource Group](https://docs.microsoft.com/en-us/azure/aks/faq#why-are-two-resource-groups-created-with-aks).
5. `NSG_RG` = Resource group where the Network Security Group is in
6. `NSG_SUB_ID` = Subscription where the Resource Group lives
7. the `EH*` values are set using the environment variables from 
   the prior setup part

Log on to your AKS cluster and create a Kubernetes secret for the six configs above:

```sh
kubectl create secret "aks-nsg-ingress-ban-ip-secrets" \
  --from-literal=AAD_ID="YOUR_AAD_ID" \
  --from-literal=SP_ID="YOUR_SP_ID" \
  --from-literal=SP_KEY="YOUR_SP_KEY" \
  --from-literal=NSG="YOUR_NSG_NAME" \
  --from-literal=NSG_RG="YOUR_NSG_RESOURCE_GRP" \
  --from-literal=NSG_SUB_ID="YOUR_NSG_SUBSCRIPTION_ID" \
  --from-literal=EH_NS="$EVENT_HUB_NAMESPACE" \
  --from-literal=EH_NAME="$LOGS_EVENT_HUB" \
  --from-literal=EH_KEY_NAME="aks-nsg-ingress-ban-ip" \
  --from-literal=EH_KEY="$EH_KEY" \
  --from-literal=HEALTH_TOKEN="SOME_KEY_STRING" \
  -n "your-namespace"
```
(replace all "YOUR..." by the real values of course)

Have a look at [run-kubernetes-pod.yaml](run-kubernetes-pod.yaml). 
This only needs one change: In the Minion Ingress change this `host: YOUR_DOMAIN` config. 
(If you don't have a [Mergeable Ingress with Master/Minion enabled controller](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/mergeable-ingress-types), you can ignore or delete these section)
Done that you are ready to run the pod. 

To start the `ks-nsg-ingress-ban-ip` pod simply run
```sh
kubectl apply run-kubernetes-pod.yaml -n "your-namespace"
```

Check the logs: It should print out something about starting, login ... and if you find:
```sh 
kubectl logs aks-nsg-ingress-ban-ip -n "your-namespace"
...
... EH: Ready ...
```
Yej :-), everything went well and the pod is waiting for NGINX ingress logs now.

Not much logs are printed out in `LOG_LEVEL: info` after startup. 
Only `"Ban IP address ..."` will be logged. 

Some other insights and metrics are available on the health web endpoint:
http://localhst:8080/aks-nsg-ingress-ban-ip/healthz?token=SOME_KEY_STRING

The `token` protects providing detailed insights. 

To toggle thesome access log printed out on console, there is a `healthz/tracelogs` endpoint:
http://localhst:8080/aks-nsg-ingress-ban-ip/healthz/tracelogs?token=SOME_KEY_STRING


To see all the activities, you can change the `LOG_LEVEL` to `debug` in the Pod config.

IMPORTANT: The pod YAML uses "latest" version of the pod, you may want to change this ;-)

# NSG Rule Details 

Blacklist security rules use "prio" of 1000...10365 (1000 + day)

## Whitelisting IPs Manually

If you need to set up whitelist rules, you should use a prio with e.g. 200 to avoid conflicts.

CHANGELOG.md
# 0.5.6
- fix log format detection
- improved log parser

# 0.5.6
- improve health response
- default token

# 0.5.5
- fix clean up

# 0.5.4
- log parser improvement (for strange hex logs)
- toggle trace logs on/off via HTTP call

# 0.5.2 
- fix bug in cleanup blacklists (sometimes too fast)

# 0.5.2 
- EH config startup log 
- improved exception handling

test/testLogs.js
const assert = require('assert')
const ehLog  = require( '../app/eh-logs' )

describe( 'Log Pattern Parser', () => {

  before(() => {
    ehLog.init(
      { msgCnt: 0, errCnt: 0, bannedIPs: {}, lastLog:[], tracelogs: true }, 
      { nogoPatterns: '' }
    )
    process.env.LOG_LEVEL = 'TRACE'
  })

  it( 'parse normal logs', () => {
    let record = {
      LogEntry: '18.14.85.143 - [18.14.85.143] - - [10/Nov/2020:07:31:20 +0000] "GET /icons.svg HTTP/2.0" 200 9068 "https://somepage.com/orders" "Mozilla/5.0 (Macintosh  Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36" 36 0.002 [frontend-80] 10.244.1.164:80 27158 0.004 200 0b354a3c3860bb5239c48a3b6f2028eb'
    }
    let logDta  = ehLog.checkAndParseAccessLog( record )
    assert.equal( logDta.code, 200 )
  })

  it( 'parse hex strange logs', () => {
    let record = {
      LogEntry: '66.240.205.34 - [66.240.205.34] - - [10/Nov/2020:08:51:55 +0000] \"H\\x00\\x00\\x00tj\\xA8\\x9E#D\\x98+\\xCA\\xF0\\xA7\\xBBl\\xC5\\x19\\xD7\\x8D\\xB6\\x18\\xEDJ\\x1En\\xC1\\xF9xu[l\\xF0E\\x1D-j\\xEC\\xD4xL\\xC9r\\xC9\\x15\\x10u\\xE0%\\x86Rtg\\x05fv\\x86]%\\xCC\\x80\\x0C\\xE8\\xCF\\xAE\\x00\\xB5\\xC0f\\xC8\\x8DD\\xC5\\x09\\xF4\" 400 157 \"-\" \"-\" 0 0.174 [] - - - - ccae4a996c58deae9dfab0c4d97d220d'
    }
    let logDta  = ehLog.checkAndParseAccessLog( record )
    assert.equal( logDta.code, 400 )
  })


  it( 'parse new log format', () => {
    let record = {
      LogEntry: '41.19.81.90 - - [14/Dec/2020:08:50:03 +0000] "GET /news/show-info HTTP/1.1" 200 85 "-" "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0; AppInsights)" 590 0.004 [misc-80] [] 10.244.18.117:80 85 0.004 200 0aa4568b92448364341c2d4e476a1b3f'
    }
    let logDta  = ehLog.checkAndParseAccessLog( record )
    assert.equal( logDta.code, 200 )
  })

})

app/health.js
const express  = require( 'express' )
const pjson    = require( './package.json' )
const log      = require( './log' ).logger

module.exports = {
  initHealthEndpoint
}

const healthPort = 8080
let   healthPath = null
let   token      = null

// ----------------------------------------------------------------------------
let expressApp  = express()

let metrics = null
let upSince = ( new Date() ).toISOString()


function initHealthEndpoint( path, healtMetrics, authToken ) {
  metrics = healtMetrics
  healthPath = path
  token      = authToken

  if ( healthPort && healthPath ) {

    expressApp.get( healthPath, (req, res) => {
      log.debug( 'GET '+healthPath, req.query )
      let resJSON = {
        name      : pjson.name,
        version   : pjson.version,
        upSince   : upSince,
        metrics   : metrics
      }
      if ( token ) {
        if ( req.query && req.query.token != token ) {
          resJSON = {}
        }
      }
      res.send( resJSON )
    })

    expressApp.get( healthPath+'/tracelogs', (req, res) => {
      log.info( 'GET '+healthPath+'/tracelogs' )
      if ( req.query && req.query.token && token ) {
        if ( req.query.token == token ) {
          metrics.tracelogs = ! metrics.tracelogs
          log.info( 'tracelogs=' +metrics.tracelogs )
        }
      } else if ( req.query.token == 'test' ) {
          metrics.tracelogs = ! metrics.tracelogs
          log.info( 'Use default token!!' )
          log.info( 'tracelogs=' +metrics.tracelogs )
      }
      res.send( { tracelogs: metrics.tracelogs } )
    })

    expressApp.listen( healthPort )
    log.info( `Health app listening on port ${healthPort} and path "${healthPath}"` )

  } else {
    log.info( 'Health check endpoint not configured' )
  }
}

app/eh-logs.js
const { EventHubClient, EventPosition } = require( '@azure/event-hubs' )
const log    = require( './log' ).logger

module.exports = {
  init,
  startEhStreamReceiver,
  checkAndParseAccessLog
}

let penaltyReduceJob = null

// ----------------------------------------------------------------------------
// Configs

let cfg    = null
let status = null

let max = {
  'nogo' : 50,
  'err' :  3
}
let nogoPatterns = []

let ehConnStr = null

// ----------------------------------------------------------------------------
// Data fields

let errorIPs = {}

let banIpFn = null

// ----------------------------------------------------------------------------
function init( statusDta, config ) {
  cfg    = config
  status = statusDta

  ehConnStr = 'Endpoint=sb://'+cfg.ehNameSpace+'.servicebus.windows.net/;'+
    'SharedAccessKeyName='+cfg.ehKeyName+';SharedAccessKey='+cfg.ehKey

  nogoPatterns = cfg.nogoPatterns.split(',')
  max[ 'nogo' ] = cfg.nogoMax
  max[ 'err' ] = cfg.errorsMax

  log.info( 'Error Thresholds ',max )
  log.info( 'NoGo Patterns',nogoPatterns )

  // reduce penalty count every minute
  penaltyReduceJob = setInterval( reducePenaltyCount, 60 * 1000 ) 
  errorIPs = status.errorIPs
}

// ----------------------------------------------------------------------------
function reducePenaltyCount() {
  for ( let ip in errorIPs ) {
    //log.debug( 'reducePenaltyCount', ip, errorIPs[ip]['nogo'] , errorIPs[ip]['err'] )
    if ( errorIPs[ip]['nogo'] > 0  ) {
      errorIPs[ip]['nogo']--
      log.debug( 'reduce nogo', ip, errorIPs[ip]['nogo'] )
    }
    if ( errorIPs[ip]['err'] > 0  ) {
      errorIPs[ip]['err']--
      log.debug( 'reduce err', ip, errorIPs[ip]['err'] )
    }
    if ( errorIPs[ip]['err'] == 0  && errorIPs[ip]['err'] == 0) {
      log.debug( 'del from list', ip  )
      delete errorIPs[ ip ]
    }
  }
}

// ----------------------------------------------------------------------------
// Receive data after processing from Azure Event Hub

async function startEhStreamReceiver( banIpCallback ) {
  try {

    log.info( 'EH: Start: '+cfg.ehNameSpace+'/'+cfg.ehName )
    const client = EventHubClient.createFromConnectionString( 
      ehConnStr, 
      cfg.ehName
    )

    log.info( 'EH: Get partitions...' )
    const allPartitionIds = await client.getPartitionIds()
    
    log.info( 'EH: Starting receivers...' )
    for ( let partition of allPartitionIds ) {
      log.info( 'EH: ... start receiver on partition '+partition )
      const rcvOpts =  { eventPosition: EventPosition.fromEnqueuedTime( Date.now() )  }
      receiveHandler = client.receive( partition, onMessage, onErr, rcvOpts )  
    }

    banIpFn = banIpCallback
    log.info( 'EH: Ready.' )

  } catch ( exc ) {
    log.error( exc )
    throw( exc )
  }
}

// ----------------------------------------------------------------------------

const onErr = ( error ) => {
  log.error( 'EH: Error when receiving message: ', error )
}

// ----------------------------------------------------------------------------

const onMessage = ( eventData ) => {
  try { 
    // eventhub delivers messages as bulk:
    for ( let record of extractLogArr( eventData ) ) {

      let needBan = false      
      let logDta  = checkAndParseAccessLog( record )
      
      // only access log with error codes:
      if ( isNogoPattern( logDta ) ) {
        needBan = addViolationForIP( logDta, 'nogo' )
      } else if ( isHttpError( logDta ) ) {
        needBan = addViolationForIP( logDta, 'err' )
      } 

      if ( needBan ) {
        log.info( 'Ban IP address: ' + logDta.ip )
        logDta.banTime = (new Date()).toISOString()
        banIpFn( logDta )
      }
    }
  } catch ( e ) {  log.error( 'EH receive', e ) }
}

// ----------------------------------------------------------------------------
// helper

function logLastEvents( record ) {
  if ( record.LogEntry ) {
    status.lastLog.push( record.LogEntry )
    if ( status.lastLog.length > 20 ) {
      status.lastLog.shift()
    }
  }
}

function addViolationForIP( logDta, reason ) {
  log.debug( reason, logDta )
  status.errCnt++
  if ( ! errorIPs[ logDta.ip ] ) {// IP caused no violations before
    errorIPs[ logDta.ip ] = {
        'nogo' : 0,
        'err' : 0
      }
  }
  errorIPs[ logDta.ip ][ reason ]++
  log.debug( 'check max', logDta.ip, errorIPs[ logDta.ip ][ reason ], max[ reason ] )
  if ( errorIPs[ logDta.ip ][ reason ] >= max[ reason ] ) {
    return true
  }
  return false
}

function isHttpError( logDta ) {
  if ( logDta && logDta.code >= 400 ) { 
    return true 
  }
  return false
}

function isNogoPattern( logDta ) {
  if ( ! logDta ) { return }
  for ( let nogo of nogoPatterns ) {
    if ( logDta.op.indexOf( nogo ) != -1 ) {
      log.info( 'NOGO!!', logDta.op, nogo )
      return true 
    } 
  }
  return false
}

function extractLogArr( eventData ) {
  if ( eventData.body && eventData.body.records ) {
    return eventData.body.records
  } else {
    return []
  }
}

function checkAndParseAccessLog( record ) {
  if ( record.LogEntry && record.LogEntry.indexOf('- - [') > 0 ) { 
    logLastEvents( record ) 

    let logSplitStr = record.LogEntry.split('"')

    let callStr  = logSplitStr[1]
    let callerIP = logSplitStr[0].split(' ')[0]
    let codeStr  = logSplitStr[2].trim().split(' ')[0]

    if ( status.tracelogs ) { 
      log.info( callerIP +' '+ callStr +' '+ codeStr )
    }

    let code = parseInt( codeStr, 10 )
    if ( isNaN( code ) ) { code = -1  }
    let result = {
      ip   : callerIP,
      op   : callStr,
      code : code
    }
    status.msgCnt++
    return result
  }
  return null
}

app/app.js
const nsg    = require( 'azure-nsg-ban-ips' )
const log    = require( './log' ).logger
const pjson  = require( './package.json' )
const ehLogs = require( './eh-logs' )
const stats  = require( './health' )

log.info( 'Starting '+pjson.name+' v'+pjson.version )

let cfg = readConfig()

let status = {
  msgCnt    : 0,
  errCnt    : 0,
  tracelogs : false,
  lastLog   : [],
  errorIPs  : {},
  bannedIPs : {}
}

run()

// run NSG clean up every hour
setInterval( unbanIPs, 60 * 60 * 1000 ) 

// ----------------------------------------------------------------------------
let banning = false 
let banBacklog = []
let bannedIPs = []

async function run() {
  try {
    log.info( 'Login for NSG operations...' )
    log.info( 'Sub: '+cfg.nsgSubId +' RG:'+cfg.nsgRG +' NSG:'+cfg.nsgName )

    // get credentials for all following operations first:
    await nsg.login( cfg.spId, cfg.spKey, cfg.aadId, cfg.nsgSubId, cfg.nsgRG, cfg.nsgName )
    
    log.info( 'Clean up NSGs first...' )
    nsg.cleanupOldBlacklists()

    log.info( 'Start EventHub listener...' )
    ehLogs.init( status, cfg )
    ehLogs.startEhStreamReceiver( banIPaddrCallback )

    stats.initHealthEndpoint( cfg.healthzPath, status, cfg.healthToken )

  } catch ( exc ) { 
    log.error( 'Exception in MAIN run()', exc ) 
    process.exit( 0 )
  }
}

// ----------------------------------------------------------------------------
// process banning in NSG

async function banIPaddrCallback( logData ) {
  try {
    let maliciousIPaddr = logData.ip
    if ( bannedIPs && bannedIPs.indexOf( maliciousIPaddr ) >=0 ) {
      return
    }
    banBacklog.push( maliciousIPaddr )
    status.bannedIPs[ maliciousIPaddr ] = logData

    // this should also handles DDoS attacks well
    if ( ! banning ) { // avoid running multiple NSG ops at the same time
      banning = true
      log.debug( 'Banning mode ON' )
      while ( banBacklog.length > 0 ) { 
        log.debug( 'banBacklog', banBacklog )
        let clonedBanList = JSON.parse( JSON.stringify( banBacklog ) ) 
        banBacklog = [] 
        try {
          log.debug( 'calling addIpAddrArrToBlacklist...' )
          bannedIPs = await nsg.addIpAddrArrToBlacklist( clonedBanList )
          log.debug( 'Blacklist', bannedIPs )
        } catch ( e ) { log.error( e ) }
      }
      banning = false
      log.debug( 'Banning mode OFF' )
    }

  } catch ( exc ) { 
    log.error( exc ) 
    banning = false
  }
}

// ----------------------------------------------------------------------------
// remove all ban rules older than 2 days
function unbanIPs() {
  nsg.cleanupOldBlacklists()
}

// ----------------------------------------------------------------------------
// helper

function readConfig() {
  let mustDie = false
  let configs = { 
    aadId        : 'AAD_ID', 
    spId         : 'SP_ID',
    spKey        : 'SP_KEY',
    nsgName      : 'NSG',
    nsgRG        : 'NSG_RG',
    nsgSubId     : 'NSG_SUB_ID',
    ehNameSpace  : 'EH_NS',
    ehName       : 'EH_NAME',
    ehKeyName    : 'EH_KEY_NAME',
    ehKey        : 'EH_KEY',
    errorsMax    : 'ERROR_THRESHOLD',
    nogoPatterns : 'NOGO_REQUESTS',
    nogoMax      : 'NOGO_THRESHOLD',
    healthzPath  : 'HEALTH_PATH'
  }
  for ( let aCfg in configs ) {
    let cfgVar = configs[ aCfg ]
    if ( process.env[ cfgVar ] ) {
      configs[ aCfg ] = process.env[ cfgVar ] 
    } else {
      console.log( 'ERROR: Environment variable '+cfgVar+' not set.' )
      mustDie = true
    }
  }
  if ( process.env[ 'HEALTH_TOKEN' ] ) {
    configs.healthToken = process.env[ 'HEALTH_TOKEN' ] 
  }
  if ( mustDie ) { 
    process.exit( 0 ) 
  }
  return configs
}